micropip/_commands/freeze.py

```diff
@@ -40,12 +40,17 @@ def freeze_data() -> dict[str, Any]:
 
 
 def load_pip_packages() -> Iterator[tuple[str, dict[str, Any]]]:
-    return filter(None, map(load_pip_package, importlib.metadata.distributions()))
+    return map(
+        package_item,
+        filter(None, map(load_pip_package, importlib.metadata.distributions())),
+    )
```

Comment:
I'm curious how you feel about the readability of `map` and `filter` in Python where we don't have any way to do right function application. What I'd really like to see is
```py
importlib.metadata.distributions().map(load_pip_package).filter(None).map(package_item)
```
so that the logic flows from left to right and top to bottom instead of in a spiral. I tend to use comprehensions instead but they are arguably just as bad if not worse. Combining filter and map does make sense but list comprehensions do it in a strange order. I do like Rust's `filter_map` quite a lot.

Anyways, code is clear I'm happy to do it this way.

Reply:
Personally, I'm a big fan of the general concept of filter and map (and reduce). Part of my affinity comes from my first formal language training being in [Scheme](https://en.wikipedia.org/wiki/Scheme_(programming_language)). But my reason for sticking with it is how it incentivizes stateless functional programming. Comprehensions and generator expressions achieve the same thing, but often by adding unnecessary variables. If I'm working with a comprehension and it has the form `[do(x) for x in input if check(x)]`, I will almost certainly translate that to `map(do, filter(check, input))` and avoid creating the name `x`. I will sometimes separate the two lines into `filtered = filter(check, input);` `map(do, filtered)` if the name `filtered` improves readability (which it could in the case above).

I do avoid `map`/`filter` when their use feels contrived or convoluted for the application.


> What I'd really like to see is
> 
> ```python
> importlib.metadata.distributions().map(load_pip_package).filter(None).map(package_item)
> ```

I agree - that reads better. And other languages (Node, Ruby, ...) design that paradigm into their language. Python, on the other hand, has historically discouraged these chains of operations. So my pragmatic approach is to accept that the language constructs are somewhat inelegant, but still preferable to the alternative of eager-evaluated `for/append` loops.

Reply:
Yeah a lot of my early programming experience was also with lisps. I particularly enjoy programming with only immutable data structures. But it's not always the best for performance.

I'm really excited about the iterator helpers spec in JavaScript which will make doing all of this lazily very ergonomic too:   
https://github.com/tc39/proposal-iterator-helpers